### PR TITLE
fix some codegen bugs: NIM_BOOL, NIM_STATIC_ASSERT, --passc:-std=... (etc)

### LINKED ISSUE #13933: Unable to override -std:gnu++14
Function `echo` outputs the wrong string.

### hello.nim
```nim
{.passc: "-std=c++17".}
{.compile: "uWebSockets/src/App.h".}
echo "Hello World!"

# nim cpp --cincludes:"uWebSockets/uSockets/src/" hello.nim
# nim cpp --passC:-std=c++17 --cincludes:"uWebSockets/uSockets/src/" hello.nim
```

### Current Output
```
work/scratch/nim/uWebSockets/src/App.h: In lambda function:
work/scratch/nim/uWebSockets/src/App.h:154:18: error: ‘string_view’ is not a member of ‘std’
  154 |             std::string_view secWebSocketKey = req->getHeader("sec-websocket-key");
      |                  ^~~~~~~~~~~
work/scratch/nim/uWebSockets/src/App.h:154:18: note: ‘std::string_view’ is only available from C++17 onwards
work/scratch/nim/uWebSockets/src/App.h:155:17: error: ‘secWebSocketKey’ was not declared in this scope
```

### Expected Output
It should compile

### Possible Solution

Something to do with this?
https://github.com/nim-lang/Nim/blob/devel/compiler/extccomp.nim#L89

### Additional Information

This is my first ticket.  Should I just compile nim with that line and file changed?

```
$ nim -v
Nim Compiler Version 1.2.0 [Linux: amd64]
Compiled at 2020-04-03
Copyright (c) 2006-2020 by Andreas Rumpf

active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by timotheecour**
* `--passc:-std=c++17` (or similar flags) are now honored instead of silently ignored, by putting `--passc` options **after** instead of **before** the default
* use correct definition for `NIM_BOOL ` in C target. Previous definition was incorrect, as `bool` is not guaranteed to be compatible with `unsigned char` (could have different sizeof etc); furthermore, it would cause codegen error for C code that uses `_Bool` or (after `include <stdbool.h>`) `bool`
note: it's >= C99 but I doubt nim codegen works with < C99; if needed an extra check (or conditional define) can be used

* `NIM_NIL` is incorrect for c++ < c++11, but fixing it would be hard and maybe not worth it, eg:
```nim
when true:
  {.emit:"""
  #include <stdio.h>
  struct Foo {
    int fun(void* a) { return 1;}
    int fun(int a){ return 2; }
  };
  """.}
  type Foo {.importcpp.} = object
  proc fun(foo: var Foo, a: pointer): cint {.importcpp.}
  var foo: Foo
  echo foo.fun(nil)
```

```
nim cpp -r --passc=-std=c++03 $timn_D/tests/nim/all/t10447.nim
2
nim cpp -r --passc=-std=c++17 $timn_D/tests/nim/all/t10447.nim
1
```
(requires rebuilding nim with this PR so that `--passc=-std=c++03` is honored)

## EDIT
* fixes https://github.com/nim-lang/Nim/issues/13933

## links
* https://stackoverflow.com/questions/1608318/is-bool-a-native-c-type/1608350
* D20200328T203812

### CODE REVIEW

File: compiler/extccomp.nim [hunk 1]
```diff
@@ -612,6 +612,11 @@ proc getCompileCFileCmd*(conf: ConfigRef; cfile: Cfile,
   # We produce files like module.nim.cpp, so the absolute Nim filename is not
   # cfile.name but `cfile.cname.changeFileExt("")`:
   var options = cFileSpecificOptions(conf, cfile.nimname, cfile.cname.changeFileExt("").string)
+  if useCpp(conf, cfile.cname):
+    # needs to be prepended so that --passc:-std=c++17 can override
+    # we could avoid allocatoin by making cFileSpecificOptions inplace
```
Comment (cooldome) [inline compiler/extccomp.nim]: typo: allocatoin

**Comment by Araq** [comment]
``stdbool.h`` is useless though as it doesn't say anything about ``sizeof(bool)`` and Nim requires it to be 1 byte.

File: compiler/extccomp.nim (hunk 1 above)
Comment (timotheecour) [reply compiler/extccomp.nim]: fixed

**Comment by timotheecour** [comment]
PTAL

> stdbool.h is useless though as it doesn't say anything about sizeof(bool) and Nim requires it to be 1 byte.

* removed `stdbool.h` and used `_Bool` directly instead to defined `NIM_BOOL`, which still fixes issue in top post
* I'm defining a portable `NIM_STATIC_ASSERT` and checking `NIM_STATIC_ASSERT(sizeof(bool)==1)`; if a weird platform shows up where `sizeof(_Bool)!=1`, we'll have a CT error instead of potentially weird bugs and we'll adjust nimbase.h accordingly (eg via `#ifdef NIM_DEFINE_BOOL_AS_UCHAR`)
*  `NIM_STATIC_ASSERT` is better (and easier to reuse) than the old hack as it shows line number and failing condition; I added a test for checking its behavior

* `NIM_STATIC_ASSERT` is a key ingredient to support CT sizeof for importc types, see https://github.com/nim-lang/RFCs/issues/205 for details

* couldn't figure out how to use the testament `discard` spec for a test like `tests/ccgbugs/mstatic_assert.nim` so I'm instead compiling it in `tests/trunner.nim` (after refactoring that file from `tests/vm/tevalffi.nim`); the few other tests that also need an explicit compilation step (or non-trivial checks on the output) can also be simplified by merging their logic into `trunner` (better than making testament spec more complex with more magic flags IMO)

* related to that, I've fixed a bug so that CTFFI now tests the c++ (instead of C) target depending on `NIM_COMPILE_TO_CPP`

**Comment by Araq** [comment]
> it's >= C99 but I doubt nim codegen works with < C99; if needed an extra check (or conditional define) can be used

We support C89.

**Comment by timotheecour** [comment]
> We support C89.

done, that was a simple fix, PTAL

File: lib/nimbase.h [hunk 2]
```diff
@@ -288,34 +303,40 @@ __AVR__
...
-#  ifndef NIM_FALSE
-#    define NIM_FALSE false
-#  endif
-#  define NIM_BOOL bool
+#define NIM_BOOL bool
+#elif (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901)
+// see #13798: to avoid conflicts for code emitting `#include <stdbool.h>`
+#define NIM_BOOL _Bool
+#else
+typedef unsigned char NIM_BOOL // best effort
```
Comment (Araq) [inline lib/nimbase.h]: Missing ';'

Comment (timotheecour) [reply lib/nimbase.h]: nice catch, fixed

**Comment by timotheecour** [comment]
friendly ping @Araq

**Pull request merged**